src/v/kafka/server/handlers/write_txn_markers.cc

```diff
@@ -0,0 +1,114 @@
+// by the Apache License, Version 2.0
+
+#include "kafka/server/handlers/write_txn_markers.h"
+
+#include "cluster/errc.h"
+#include "cluster/partition_manager.h"
+#include "cluster/shard_table.h"
+#include "cluster/tx_gateway_frontend.h"
+#include "cluster/types.h"
+#include "kafka/protocol/errors.h"
+#include "kafka/protocol/schemata/write_txn_markers_request.h"
+#include "kafka/protocol/schemata/write_txn_markers_response.h"
+#include "model/record.h"
+
+#include <seastar/core/coroutine.hh>
+
+namespace kafka {
+
+namespace {
+
+ss::future<std::error_code> expire_tx_for_ntp(
+  request_context& ctx, const model::ntp& ntp, model::producer_identity pid) {
+    auto& shard_table = ctx.connection()->server().shard_table();
+    auto shard = shard_table.shard_for(ntp);
+
+    if (!shard) {
+        return ss::make_ready_future<std::error_code>(
+          cluster::tx_errc::shard_not_found);
+    }
+
+    auto& partition_manager = ctx.connection()->server().partition_manager();
+
+    return partition_manager.invoke_on(
+      *shard,
+      [pid, ntp](cluster::partition_manager& mgr) mutable
+      -> ss::future<std::error_code> {
+          auto partition = mgr.get(ntp);
+          if (!partition) {
+              return ss::make_ready_future<std::error_code>(
+                cluster::tx_errc::partition_not_found);
+          }
+
+          auto stm = partition->rm_stm();
+
+          if (!stm) {
+              vlog(
+                cluster::txlog.warn,
+                "can't get tx stm of the {}' partition",
+                ntp);
+              return ss::make_ready_future<std::error_code>(
+                cluster::tx_errc::stm_not_found);
+          }
+
+          return stm->mark_expired(pid);
+      });
+}
+
+} // namespace
+
+template<>
+ss::future<response_ptr> write_txn_markers_handler::handle(
+  request_context ctx, [[maybe_unused]] ss::smp_service_group g) {
+    write_txn_markers_request request;
+    request.decode(ctx.reader(), ctx.header().version);
+    vlog(klog.trace, "Handling request {}", request);
+
+    write_txn_markers_response resp;
+
+    for (const auto& marker : request.data.markers) {
+        model::producer_identity pid;
+        pid.id = marker.producer_id;
+        pid.epoch = marker.producer_epoch;
+
+        writable_txn_marker_result marker_res;
+        marker_res.producer_id = marker.producer_id;
+        for (const auto& topic : marker.topics) {
+            writable_txn_marker_topic_result topic_res;
+            topic_res.name = topic.name;
+            for (const auto& partition : topic.partition_indexes) {
+                model::ntp ntp(model::kafka_namespace, topic.name, partition);
+                auto res = co_await expire_tx_for_ntp(ctx, ntp, pid);
+                writable_txn_marker_partition_result partition_res;
+                partition_res.partition_index = partition;
+                switch (res.value()) {
+                case static_cast<int>(cluster::tx_errc::none):
```

Comment:
this might be an indication that `expire_tx_for_ntp` should return a different error type?